File: 07_valid_parameters.ts
function isValid(s: string): boolean {
    if (s.length === 1) return false
    
    let stack = []
    let c = 0

    for (let i = 0; i < s.length; i++) {
        // const element = array[i]
        if (
            s[i] === "(" ||
            s[i] === "[" ||
            s[i] === "{" 
        ) {
            stack.push(s[i])
            c++
            continue
        }

        if (
            (stack.at(-1) === "(" && s[i] === ")") ||
            (stack.at(-1) === "[" && s[i] === "]") ||
            (stack.at(-1) === "{" && s[i] === "}")
        ) {
            stack.pop()
            continue
        } else {
            return false
        }
    }

    return stack.length > 0 ? false : true
};

// const result_07 = isValid("[()]")  // false
// const result_07 = isValid("[")     // false
// const result_07 = isValid("[[")    // false
// const result_07 = isValid("([]){") // false
const result_07 = isValid("([)]")    // false
console.log("result: ", result_07)

File: 07_valid_parameters.rs
struct Solution {}

impl Solution {
    // pub fn is_valid(s: &str) -> bool { // not code changes when using &str
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = vec![];

        for i in 0..s.len() {
            let current = s.chars().nth(i).unwrap();
            
            if stack.len() > 0 {
                let last = stack.last().unwrap();

                if Self::is_pair(*last, current) {
                    stack.pop();
                    continue;
                }
            }

            stack.push(current);
        }

        return stack.len() == 0;
    }


    pub fn is_pair(last: char, current: char) -> bool {
        (last == '(' && current == ')') ||
        (last == '{' && current == '}') ||
        (last == '[' && current == ']')
    }
}

fn main() {
    let result: bool = Solution::is_valid(("[()]").to_string());      // true
    // let result: bool = Solution::isValid("[()]")     // false
    // let result: bool = Solution::isValid("[")        // false
    // let result: bool = Solution::isValid("[[")       // false
    // let result: bool = Solution::isValid("([]){")    // false
    // let result: bool = Solution::isValid("([)]")     // false
    println!("{}", result);
}

// notes
// line 11: take care to move out the if, make sure values exists inside first
// thats why we need to avoid unwrap() in production.